File: backend/schemas/root_cause.py
"""Root cause analysis schemas (T2.4 minimal closed loop)."""

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field

# ---------------------------------------------------------------------------
# LLM output contract
# ---------------------------------------------------------------------------


class RootCauseLLMResult(BaseModel):
    """Schema the LLM must return for a root cause analysis.

    Deliberately loose (dicts passthrough) so prompt-format drift does not
    break parsing; the CoT structure lives in ``reasoning_steps`` and
    ``evidence_chain`` as free dictionaries.
    """

    model_config = ConfigDict(protected_namespaces=(), extra="ignore")

    root_cause_category: str = Field(
        description="misconfiguration | attack | failure | human_error | false_positive | unknown"
    )
    root_cause_subcategory: str | None = None
    confidence: float = Field(ge=0.0, le=1.0, default=0.5)
    reasoning_steps: list[dict] = Field(default_factory=list)
    evidence_chain: list[dict] = Field(default_factory=list)
    verification_steps: list[str] = Field(default_factory=list)
    suggested_remediation: str | None = None
    remediation_priority: str = "medium"


# ---------------------------------------------------------------------------
# API payloads
# ---------------------------------------------------------------------------


class RootCauseAnalysisResponse(BaseModel):
    """A persisted root cause analysis record."""

    model_config = ConfigDict(from_attributes=True)

    id: str
    alert_id: str
    root_cause_category: str
    root_cause_subcategory: str | None
    confidence: float
    reasoning_steps: list[dict]
    evidence_chain: list[dict]
    verification_steps: list[str]
    suggested_remediation: str | None
    remediation_priority: str
    ai_model: str
    analysis_duration_ms: float | None
    human_verified: bool
    human_feedback: str | None
    feedback_category: str | None
    created_at: str


class RootCauseAnalysisListResponse(BaseModel):
    items: list[RootCauseAnalysisResponse]
    total: int


class RootCauseFeedbackRequest(BaseModel):
    verdict: Literal["accurate", "partially_accurate", "inaccurate"]
    comment: str | None = Field(None, max_length=2000)

File: frontend/lib/api/rootCause.ts
/**
 * Root Cause Analysis API (T2.4)
 *
 * Endpoints:
 * - POST /api/v1/alerts/{id}/root-cause-analysis   → run analysis (refuses when LLM degraded)
 * - GET  /api/v1/alerts/{id}/root-cause-analyses   → analysis history for the alert
 * - POST /api/v1/root-cause-analyses/{id}/feedback → analyst verdict
 *
 * 契约对应 backend/schemas/root_cause.py。
 */

import { apiClient } from "./client";

// ── Types ────────────────────────────────────────────────

export interface RootCauseEvidence {
  evidence?: string;
  supports?: string;
  strength?: string;
  [key: string]: unknown;
}

export interface RootCauseReasoningStep {
  step?: number;
  description?: string;
  findings?: string[];
  [key: string]: unknown;
}

export type RootCauseFeedbackVerdict = "accurate" | "partially_accurate" | "inaccurate";

export interface RootCauseAnalysis {
  id: string;
  alert_id: string;
  root_cause_category: string;
  root_cause_subcategory: string | null;
  confidence: number;
  reasoning_steps: RootCauseReasoningStep[];
  evidence_chain: RootCauseEvidence[];
  verification_steps: string[];
  suggested_remediation: string | null;
  remediation_priority: string;
  ai_model: string;
  analysis_duration_ms: number | null;
  human_verified: boolean;
  human_feedback: string | null;
  feedback_category: RootCauseFeedbackVerdict | null;
  created_at: string;
}

export interface RootCauseAnalysisList {
  items: RootCauseAnalysis[];
  total: number;
}

// ── API ──────────────────────────────────────────────────

export async function runRootCauseAnalysis(alertId: number | string): Promise<RootCauseAnalysis> {
  return apiClient.post<RootCauseAnalysis>(`/api/v1/alerts/${alertId}/root-cause-analysis`);
}

export async function listRootCauseAnalyses(
  alertId: number | string,
  limit = 20
): Promise<RootCauseAnalysisList> {
  return apiClient.get<RootCauseAnalysisList>(
    `/api/v1/alerts/${alertId}/root-cause-analyses?limit=${limit}`
  );
}

export async function submitRootCauseFeedback(
  rcaId: string,
  verdict: RootCauseFeedbackVerdict,
  comment?: string
): Promise<RootCauseAnalysis> {
  return apiClient.post<RootCauseAnalysis>(`/api/v1/root-cause-analyses/${rcaId}/feedback`, {
    verdict,
    comment: comment ?? null,
  });
}
